lib/spack/spack/util/file_cache.py

```diff
@@ -117,6 +117,10 @@ def write_transaction(self, key):
         moves the file into place on top of the old file atomically.
 
         """
+        # TODO: this nested context manager adds a lot of complexity and
+        # TODO: is pretty hard to reason about in llnl.util.lock. At some
+        # TODO: point we should just replace it with functions and simplify
+        # TODO: the locking code.
```

Comment:
Just a comment for further discussion later.  The `release_fn` is to my understanding a callable that gets executed both on normal and exceptional paths when we are exiting the last nested lock section (i.e. something that releases an acquired resource no matter how we exit). I wonder if the implementation of the release semantic can be simplified further by trying to use `contextlib.contextmanager` and writing a generator with the usual `try/except/finally` block.